### PR TITLE
Implement storage for Wards and whole lot of other stuff

### PR DESCRIPTION
**Pull request by daytona65**
Implemented storage, ward refreshing and respective wards update when their patients are updated.
30 failing testcases and I have no idea why, all throwing duplicate patient exception, suspect the typical patients data have issue but CANNOT seem to find out why

### CODE REVIEW

File: src/main/java/seedu/address/model/AddressBook.java [hunk 1]
```diff
@@ -70,36 +72,6 @@ public void resetData(ReadOnlyAddressBook newData) {
...
-                    ward.removePatient(patient);
-                    break;
-                }
-            }
-            if (!wardExists) {
-                Ward newWard = new Ward(wardName);
-                newWard.addPatient(patient);
-                addWard(newWard);
-            }
-        }
```
Comment (daytona65) [inline src/main/java/seedu/address/model/AddressBook.java:102]: Removed this ugly bit of code from previous PR as implementing Storage settles this issue

File: src/main/java/seedu/address/model/AddressBook.java [hunk 2]
```diff
@@ -127,7 +99,11 @@ public boolean hasPatientNric(Patient patient) {
      * The patient must not already exist in the address book.
      */
     public void addPatient(Patient p) {
+        if (!wards.contains(p.getWard())) { //If wardlist does not contain patient's ward, don't add it in.
+            throw new WardNotFoundException();
+        }
         patients.add(p);
+        wards.addPatient(p);
```
Comment (daytona65) [inline src/main/java/seedu/address/model/AddressBook.java:106]: Wards are edited with the patient

File: src/main/java/seedu/address/model/AddressBook.java [hunk 3]
```diff
@@ -138,16 +114,19 @@ public void addPatient(Patient p) {
...
 
     /**
      * Removes {@code key} from this {@code AddressBook}.
      * {@code key} must exist in the address book.
      */
     public void removePatient(Patient key) {
+        requireNonNull(key);
         patients.remove(key);
+        wards.remove(key);
     }
```
Comment (daytona65) [inline src/main/java/seedu/address/model/AddressBook.java:130]: Wards are edited with the patient

File: src/main/java/seedu/address/model/patient/Patient.java [hunk 4]
```diff
@@ -141,13 +141,7 @@ public void setDischarge(Discharge newDischarge) {
...
      */
     public boolean isSamePatient(Patient otherPatient) {
-        if (otherPatient == this) {
-            return true;
-        }
-
-        return otherPatient != null
-            && otherPatient.getNric().equals(this.getNric())
-            && otherPatient.getName().equals(this.getName());
+        return this.equals(otherPatient);
```
Comment (daytona65) [inline src/main/java/seedu/address/model/patient/Patient.java:144]: Redundant code

File: src/main/java/seedu/address/model/patient/Patient.java [hunk 5]
```diff
@@ -180,9 +174,7 @@ public boolean equals(Object other) {
 
         Patient otherPatient = (Patient) other;
         return otherPatient.getNric().equals(getNric())
-            && otherPatient.getName().equals(getName())
-            && otherPatient.getStatus().equals(getStatus())
-            && otherPatient.getWard().equals(getWard());
+            && otherPatient.getName().equals(getName());
```
Comment (daytona65) [inline src/main/java/seedu/address/model/patient/Patient.java:177]: Patient equals only if NRIC and NAME equal

File: src/main/java/seedu/address/model/ward/UniqueWardList.java [hunk 6]
```diff
@@ -49,25 +51,53 @@ public int size() {
...
+    /**
+     * Returns true if the list contains an equivalent ward as the given
+     * {@code String}.
+     */
+    public boolean contains(String toCheckName) {
+        requireNonNull(toCheckName);
+        Ward toCheck = new Ward(toCheckName);
+        return internalList.stream().anyMatch(toCheck::isSameWard);
+    }
+
```
Comment (daytona65) [inline src/main/java/seedu/address/model/ward/UniqueWardList.java:78]: Check for ward with string input

File: src/main/java/seedu/address/model/ward/UniqueWardList.java [hunk 7]
```diff
@@ -49,25 +51,53 @@ public int size() {
...
+     * Adds patient p to their assigned ward.
+     * @param p
+     */
+    public void addPatient(Patient p) {
+        requireNonNull(p);
+        String targetName = p.getWard();
+        int index = internalList.indexOf(wardWithName(targetName));
+        Ward target = internalList.get(index);
+        target.addPatient(p);
+        internalList.set(index, target);
```
Comment (daytona65) [inline src/main/java/seedu/address/model/ward/UniqueWardList.java:100]: Adding patient to ward too

File: src/main/java/seedu/address/model/ward/UniqueWardList.java [hunk 8]
```diff
@@ -91,6 +121,40 @@ public void setWard(Ward target, Ward editedWard) {
...
+     */
+    public void changePatientWard(Patient target, int from, int to) {
+        Ward start = internalList.get(from);
+        Ward end = internalList.get(to);
+        start.removePatient(target);
+        end.addPatient(target);
+        internalList.set(from, start);
+        internalList.set(to, end);
+    }
+
```
Comment (daytona65) [inline src/main/java/seedu/address/model/ward/UniqueWardList.java:157]: Edit patient within the ward OR move patient from ward to ward

File: src/main/java/seedu/address/model/ward/UniqueWardList.java [hunk 9]
```diff
@@ -102,6 +166,18 @@ public void remove(Ward toRemove) {
...
+     * Removes the equivalent patient from their assigned ward.
+     */
+    public void remove(Patient toRemove) {
+        requireNonNull(toRemove);
+        String targetName = toRemove.getWard();
+        int index = internalList.indexOf(wardWithName(targetName));
+        Ward targetWard = internalList.get(index);
+        targetWard.removePatient(toRemove);
+        internalList.set(index, targetWard);
+    }
```
Comment (daytona65) [inline src/main/java/seedu/address/model/ward/UniqueWardList.java:179]: remove patient from ward too

File: src/main/java/seedu/address/model/ward/Ward.java [hunk 10]
```diff
@@ -54,11 +54,20 @@ public Ward(String name) {
...
 
+    /**
+     * Ward constructor with string for comparisons
+     * @param name
+     * @return placeholder Ward
+     */
+    public static Ward wardWithName(String name) {
+        return new Ward(name);
+    }
+
```
Comment (daytona65) [inline src/main/java/seedu/address/model/ward/Ward.java:70]: Get ward with the name ID

File: src/main/java/seedu/address/storage/JsonAdaptedWard.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        }
+        if (!Capacity.isValidCapacity(capacity)) {
+            throw new IllegalValueException(Capacity.MESSAGE_CONSTRAINTS);
+        }
+        final Capacity modelCapacity = new Capacity(capacity);
+
+        return new Ward(modelName, modelCapacity);
+    }
+
+}
```
Comment (daytona65) [inline src/main/java/seedu/address/storage/JsonAdaptedWard.java:70]: Storage class for ward

File: src/main/java/seedu/address/model/ward/Ward.java (hunk 10 above)
Comment (nramapurath) [reply src/main/java/seedu/address/model/ward/Ward.java:71]: I understand this is used to find target wards in ward list operations, relying on the `equals()` method of ward checking for equality based on ward name. Good fix, but we have to be careful to not modify Ward's `equals()` function from now on.

**Pull request merged**